### DISCUSSION TITLE
Generating multiple clients in one project

### DISCUSSION THREAD

**Discussion by TobiasBales** [ExpediaGroup/graphql-kotlin#939] [Q&A] [answered]
I am trying to build a form of backend for frontend/api gateway as a proof of concept that communicates with multiple graphql services. I do want to transform the schema though so I cannot only rely on federation.
It seems like it is not possible to have more than one task generating clients (based off of multiple remote schemas).

Is this the case?
And are there ways of making this work?

**Comment by dariuszkuc** [comment, marked as answer]
Hello 👋 
If you are using Gradle it should be possible to generate multiple clients from different schemas. It might be possible to do in Maven by forcing specific phases to order the tasks but unsure if it will work.

It is pretty ugly as that was not one of our use cases but following should work

```kotlin
// configuration to generate first client against gscheibel/graphql-kotlin-demo
graphql {
    client {
        sdlEndpoint = "http://localhost:8080/sdl"
        packageName = "com.example.generated"
        clientType = GraphQLClientType.WEBCLIENT
        queryFileDirectory = "${project.projectDir.absolutePath}/src/main/resources/first"
    }
}

// above graphql extension configuration is equivalent to the below
//val graphqlDownloadSDL by tasks.getting(GraphQLDownloadSDLTask::class) {
//    endpoint.set("http://localhost:8080/sdl")
//}
//
//val graphqlGenerateClient by tasks.getting(GraphQLGenerateClientTask::class) {
//    clientType.set(GraphQLClientType.WEBCLIENT)
//    packageName.set("com.example.generated")
//    schemaFile.set(graphqlDownloadSDL.outputFile)
//    queryFileDirectory.set("${project.projectDir.absolutePath}/src/main/resources/first")
//    dependsOn("graphqlDownloadSDL")
//}

// configuration for the second client running against graphql-kotlin/examples/spring
val graphqlDownloadOtherSDL by tasks.registering(GraphQLDownloadSDLTask::class) {
    endpoint.set("http://localhost:8081/sdl")
    // explicitly define dependency on the other client generation so it will download other schema AFTER the first generation
    dependsOn("graphqlGenerateClient")
}

val graphqlGenerateOtherClient by tasks.registering(GraphQLGenerateClientTask::class) {
    clientType.set(GraphQLClientType.WEBCLIENT)
    packageName.set("com.example.generated2")
    schemaFile.set(graphqlDownloadOtherSDL.get().outputFile)
    queryFileDirectory.set("${project.projectDir.absolutePath}/src/main/resources/second")
    outputDirectory.set(project.layout.buildDirectory.dir("generated/source/graphql/main"))
    dependsOn("graphqlDownloadOtherSDL")
}

tasks {
    // original client generation task will automatically add itself to compileKotlin dependency
    // make sure that before we run compile task it will generate other client as well
    compileKotlin {
        dependsOn("graphqlGenerateOtherClient")
    }
}
```

I tested it with an example project https://github.com/dariuszkuc/graphql-kotlin-webclient-demo/commit/168646049a0dc96fc8cbacf38ba35d78dc101242 using https://github.com/gscheibel/graphql-kotlin-demo as the first server running on port 8080 and [Spring example app from graphql-kotlin repo](https://github.com/ExpediaGroup/graphql-kotlin/tree/master/examples/spring) running on port 8081.

**Reply by TobiasBales** [reply]
Awesome, thanks!
I tried using the introspection tasks but couldn't get it to work (it does not allow one to set the output file for one).

Will give it a shot and report back - probably tomorrow.

**Reply by dariuszkuc** [reply]
Yeah currently you cannot overwrite the output file names for introspection/sdl tasks -> thats why I'm explicitly forcing out the order in the above.

**Reply by TobiasBales** [reply]
Would you be open to having them overrideable? If so I would be willing to submit a PR to address it

**Reply by dariuszkuc** [reply]
Yeah I think we can make it configurable (for both Gradle and Maven plugin). Feel free to open up a PR.

**Reply by TobiasBales** [reply]
Worked great btw, thanks!
